webapp/app.py

```diff
@@ -73,11 +73,25 @@ def thank_you():
 
 @open_id.after_login
 def after_login(resp):
-    flask.session["macaroon_discharge"] = resp.extensions["macaroon"].discharge
-    flask.session["openid"] = {
-        "identity_url": resp.identity_url,
-        "email": resp.email,
+    root = flask.session["macaroon_root"]
+    discharge = resp.extensions["macaroon"].discharge
+    authorization_code = f"root={root} discharge={discharge}"
+    invitation_code = flask.session["invitation_code"]
+    data = {
+        "provider": "usso",
+        "authorization_code": authorization_code,
+        "invitation_code": invitation_code,
     }
+    response = requests.post(url=f"{ANBOXCLOUD_API_BASE}1.0/login", json=data,)
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        error = e.response.json()
+        return flask.render_template("401.html", error=error["error"])
```

Comment:
Why are you catching this?

Firstly, you're catching *all* `HTTPError`s, not just 401, so this will show a 401 for anything that might have gone wrong, which is definitely incorrect. If you wanted to do this you would do it like this:

``` python3
    except requests.exceptions.HTTPError as http_error:
        if http_error.response.status_code == 401:
            return flask.render_template(
                "401.html", error=http_error.response.json()["error"]
            )
```

But you should only look for a 401 here if you actually expect this to happen in the expected user flow. And, if I understand this flow correctly, we would never expect unauthorized at this point. If we have a macaroon and a discharge, we would expect them to be valid.

And so, if they're not valid, I'd expect to see a real error - one that would show up in sentry and alert us something went wrong with our app.

Reply:
In this case, no. I specifically tested with a new SSO user and if invitation_code POSTed here is invalid, it will return a 401, so we need to inform the user about this error (which is returned by the API). But might be right in other areas of the code (I am thinking that probably we don't need that for the first `/token` call)

Reply:
``` python3
if response.status_code == 401:
    flask.abort(401, response.json()["error"])

response.raise_for_status()
```

then

Reply:
The repetition of this 401 block does suggest that it might be worth abstracting into its own function, as @tbille suggested.

Maybe we should do something like:

``` python3
session = requests.Session()
API_BASE = "https://staging.demo-api.anbox-cloud.io/1.0/"


def api_request(url_path, method="GET", params=None, json=None, headers=None):
    """
    Make API calls to the anbox API, passing any 401 errors to Flask to handle
    """

    response = session.request(
        method,
        f"{API_BASE}/url_path",
        params=params,
        json=json,
        header=headers,
    )

    if response.status_code == 401:
        flask.abort(401, response.json()["error"])

    response.raise_for_status()

    return response
```